torch/testing/_internal/common_utils.py

```diff
@@ -691,12 +689,121 @@ def read_and_process():
             "Test is disabled because an issue exists disabling it: {}".format(disabled_test_from_issues[test_name]) +
             " To enable set the environment variable PYTORCH_RUN_DISABLED_TESTS=1")
 
+# Helper that returns the maximum value a dtype can represent
+def _get_width(dtype):
+    if dtype is torch.complex128:
+        return torch.finfo(torch.float64).bits
+    elif dtype is torch.complex64:
+        return torch.finfo(torch.float32).bits
+    elif dtype is torch.bool:
+        return 1
+    elif is_integral(dtype):
+        return torch.iinfo(dtype).bits
+    return torch.finfo(dtype).bits
+
+# Returns a dtype suitable for comparing tensors a and b.
+# NOTE: Dtypes for Comparing Tensors in the Test Framework
+# When two tensors are compared they may be of different dtypes.
+# In PyTorch, the typical type promotion rules are biased towards float
+# types, so tensors with dtypes float16 and int64 produce tensors of
+# dtype float16 when added together.
+# If we used typical PyTorch type promotion to determine the dtype that tensors
+# should be compared in, however, then the test framework would assert
+# that tensors which are actually very different were "equal" to each other.
+# For example, torch.tensor((50000), dtype=torch.long) would be "equal" to
+# torch.tensor(49984., dtype=torch.float16), regardless of the rtol and atol.
+# These cases are admittedly, unlikely, but the test framework is not intended
+# to reflect only "likely" scenarios.
+# To preserve numerical accuracy, instead of following PyTorch's type
+# promotion rules this function follows NumPy's, which are intended to preserve
+# numerics. It special-cases bfloat16 since NumPy doesn't support bfloat16.
+# TODO: if we take a NumPy dependency then this function can be greatly
+# simplified.
+def get_comparison_dtype(a, b):
+    class _TypeMeta:
+        __slots__ = ['rank', 'kind', 'width']
+
+        def __init__(self, rank, kind, width):
+            self.rank = rank
+            self.kind = kind
+            self.width = width
+
+    # Maps torch dtypes to their rank, kind and width
+    type_meta = {
+        torch.complex128 : _TypeMeta(11, 'complex', 64),
+        torch.complex64  : _TypeMeta(10, 'complex', 32),
+        torch.float64    : _TypeMeta(9, 'float', 64),
+        torch.float32    : _TypeMeta(8, 'float', 32),
+        torch.bfloat16   : _TypeMeta(7, 'float', 16),
+        torch.float16    : _TypeMeta(6, 'float', 16),
```

Comment:
it still not really clear to me what "rank" signifies semantically, or why it's required. If it's an ordering of tolerance, bfloat16 is generally less tolerant than float16.

Reply:
The ranking of bfloat16 and float16 is a little arbitrary since their comparisons are special cased. I don't know if there's too much to read into this particular private dict's semantics. Rank is just helpful for implementing the function. If it's especially irksome then the rank can be extracted out to be a ranking of the "kinds" (complex, float, int, uint) and then this table would map to the kind and bitwidth and the function could be modified to perform an additional lookup to determine which kind took precedence. Some checks (like the bfloat16 x float16 check below) would change but it's not a big deal. I just preferred one table over two.

Reply:
Per later comment: will replace this function with a table.